reconcile/utils/membershipsources/resolver.py

```diff
@@ -0,0 +1,106 @@
+from collections.abc import (
+    Iterable,
+    Sequence,
+)
+from dataclasses import dataclass
+
+from sretoolbox.utils import threaded
+
+from reconcile.gql_definitions.fragments.membership_source import (
+    AppInterfaceMembershipProviderSourceV1,
+    MembershipProviderSourceV1,
+    MembershipProviderV1,
+)
+from reconcile.utils.membershipsources.app_interface_resolver import (
+    resolve_app_interface_membership_source,
+)
+from reconcile.utils.membershipsources.models import (
+    ProviderResolver,
+    RoleBot,
+    RoleMember,
+    RoleUser,
+    RoleWithMemberships,
+)
+
+
+@dataclass
+class GroupResolverJob:
+    provider: MembershipProviderV1
+    groups: set[str]
+
+
+def build_resolver_jobs(
+    roles: Sequence[RoleWithMemberships],
+) -> Iterable[GroupResolverJob]:
+    """
+    Bundles groups to resolve by provider so that they can be resolved
+    in batches.
+    """
+    resolver_jobs: dict[str, GroupResolverJob] = {}
+    for r in roles:
+        for ms in r.member_sources or []:
+            job = resolver_jobs.get(ms.provider.name)
+            if not job:
+                job = GroupResolverJob(provider=ms.provider, groups=set())
+                resolver_jobs[ms.provider.name] = job
+            job.groups.add(ms.group)
+    return resolver_jobs.values()
```

Comment:
group by is so commonly used in this repo, but all repeated in different styles, due to `itertools.groupby` is not easy to use (need to sort first), and some objects can't be used as dict key because no `__hash__` defined, how about add a new generic util

```python
def group_by(iterable, key):
    groups = defaultdict(list)
    for item in iterable:
        groups[key(item)].append(item)
    return groups
```

Then we can convert groups to the shape we want like

```python
    groups = group_by(
        (ms for r in roles for ms in r.member_sources or []),
        key=lambda ms: ms.provider.name,
    )

    return [
        GroupResolverJob(
            provider=g[0].provider,
            groups={ms.group for ms in g},
        )
        for g in groups.values()
    ]
```

btw the return type should use concrete ones like `list` for easy covariance.

Reply:
good idea. i added such a `group_by` function in `reconcile.utils.grouping`